From 4e3b3fe8fe79d24e9b7253ed2b1167cf81cfe7c5 Mon Sep 17 00:00:00 2001
From: Andrea Arcangeli <aarcange@redhat.com>
Date: Tue, 9 Sep 2014 01:16:29 -0400
Subject: [mm] thp: turn compound_head() into BUG_ON(!PageTail) in get_huge_page_tail()

Message-id: <1410225399-8094-13-git-send-email-aarcange@redhat.com>
Patchwork-id: 93369
O-Subject: [RHEL7.1 PATCH 12/22] mm: thp: turn compound_head() into BUG_ON(!PageTail) in get_huge_page_tail()
Bugzilla: 1135506
RH-Acked-by: Rik van Riel <riel@redhat.com>
RH-Acked-by: Larry Woodman <lwoodman@redhat.com>
RH-Acked-by: Rafael Aquini <aquini@redhat.com>

From: Oleg Nesterov <oleg@redhat.com>

get_huge_page_tail()->compound_head() looks confusing.  Every caller
must check PageTail(page), otherwise atomic_inc(&page->_mapcount) is
simply wrong if this page is compound-trans-head.

Signed-off-by: Oleg Nesterov <oleg@redhat.com>
Cc: Thomas Gleixner <tglx@linutronix.de>
Cc: Dave Jones <davej@redhat.com>
Cc: Darren Hart <dvhart@linux.intel.com>
Cc: Peter Zijlstra <peterz@infradead.org>
Cc: Mel Gorman <mgorman@suse.de>
Acked-by: Andrea Arcangeli <aarcange@redhat.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Andrea Arcangeli <aarcange@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/mm.h b/include/linux/mm.h
index 21dd32a..a8ddf4b 100644
--- a/include/linux/mm.h
+++ b/include/linux/mm.h
@@ -439,13 +439,12 @@ static inline bool compound_tail_refcounted(struct page *page)
 static inline void get_huge_page_tail(struct page *page)
 {
  /*
-  * __split_huge_page_refcount() cannot run
-  * from under us.
-  * In turn no need of compound_trans_head here.
+  * __split_huge_page_refcount() cannot run from under us.
   */
+ VM_BUG_ON(!PageTail(page));
  VM_BUG_ON(page_mapcount(page) < 0);
  VM_BUG_ON(atomic_read(&page->_count) != 0);
- if (compound_tail_refcounted(compound_head(page)))
+ if (compound_tail_refcounted(page->first_page))
   atomic_inc(&page->_mapcount);
 }
 
-- 
1.7.1